feat(router): add migration for ActivatedRouteSnapshot.fragment

crisbeto opened the pull request:
Adds a migration that casts the value of `ActivatedRouteSnapshot.fragment` to be non-nullable.

Also moves some code from the `AbstractControl.parent` migration so that it can be reused.

Relates to #37336.

atscott commented on packages/core/schematics/migrations/router-fragment/util.ts:
```diff
@@ -0,0 +1,90 @@
+      const fileName = sourceFile ? basename(sourceFile.fileName) : null;
+      // Note that this `basename` check is somewhat vague, but the problem is that we can't
+      // guarantee a `node_modules` path containing `@angular/router` when running in Bazel.
+      return (fileName === 'router.d.ts' || fileName === 'router.ts') &&
```
In g3, this needed to be `router_state.d.ts` to work. Not sure if that applies to the external world as well.

crisbeto replied:
This part felt a bit flimsy to begin with. It's here primarily so we don't change something that's called `ActivatedRoute` within the user's project, but the name is specific enough that it might not matter. What do you think about removing it altogether?

atscott replied:
Given how unlikely and rare I think it would be to have something named exactly `ActivatedRoute` or `ActivatedRouteSnapshot` and also have a `fragment` property on them (unless someone is forking `@angular/router`, which still wouldn't be caught here) I think removing it is a good idea.

crisbeto replied:
Done.

crisbeto committed:
feat(router): add migration for ActivatedRouteSnapshot.fragment

Adds a migration that casts the value of `ActivatedRouteSnapshot.fragment` to be non-nullable.

Also moves some code from the `AbstractControl.parent` migration so that it can be reused.

Relates to #37336.

crisbeto commented:
@atscott I've reworked the migration so that it only applies to `ActivatedRouteSnapshot.fragment`.

josephperrott reviewed (approved):
LGTM

Reviewed-for: dev-infra
